docs: Create a new blog post for `/usr/etc/` to `/etc/` (#65)
blue-build/website: src/content/docs/blog/usretc-to-etc-transition.mdx

---

> title: Transitioning from `/usr/etc/` to `/etc/` in build-time
> description: Why we're recommending the transition from /usr/etc/ to /etc/ in build-time & instructions to do it.

Reviewer: ```suggestion
title: Place your system configuration files in `/etc/` 
description: Why we're no longer recommending writing to /usr/etc/ in custom images
```

date: 2024-08-08
authors:
    - name: fiftydinar
      img: https://avatars.githubusercontent.com/u/65243233
      github: fiftydinar
---

Hello BlueBuilders!

> In this blog post, it will be explained on why the usage of `/etc/` is recommended over `/usr/etc/` in build-time.

Reviewer: Using "I" feels more natural in a blog post, which has a more personal tone.

```suggestion
In this blog post, I will explain why the usage of `/etc/` is now recommended over `/usr/etc/` when building custom images of atomic Fedora.
```

> It is often recommended in Fedora Atomic space that the system files should be placed in `/usr/etc/`, while files in `/etc/` are reserved for local-users to make edits.
> 
> While this is true in run-time, this is technically not true in build-time (emphasis on technically).
> 
> It is a recent discovery of [Universal Blue folks](https://github.com/ublue-os) & [Rechunk developer Antheas](https://github.com/antheas) that `ostree` has different behavior regarding `/usr/etc/` & `/etc/` relationship in build-time.  
> Also thanks to the [RedHat employee & `ostree` maintainer Colin Walters](https://github.com/cgwalters) who directly notified them of this.

Reviewer: ```suggestion
We, and many others in the community have recommended previously that the system configuration files in custom images should be placed in `/usr/etc/`, while files in `/etc/` are reserved for local-users. While it is true that in run-time the `/usr/etc/` directory contains the original configuration of the image, this is technically undefined behavior build-time.

Technically, files written to `/etc/` in the custom image are what populates `/usr/etc/` on a system rebased to that image. 

It is thanks to a recent discovery by the [Universal Blue folks](https://github.com/ublue-os) & the [Rechunk](https://github.com/hhd-dev/rechunk) developer [Antheas](https://github.com/antheas) that `ostree` has different behavior regarding `/usr/etc/` & `/etc/` in build-time vs in run-time.  
Thanks to the [RedHat employee & `ostree` maintainer Colin Walters](https://github.com/cgwalters) who directly notified them of this.
```

## Why to transition

> When files are placed directly to `/usr/etc/` in build-time, no issues are noted & everything works fine for now.

Reviewer: ```suggestion
When files are placed to `/usr/etc/` in build-time, currently there are no issues & everything works as expected *for now*. 
```

> However, it is still not the ideal thing to do, since we are relying on undefined `rpm-ostree` behavior in this case, which can stop working anytime.

Reviewer: ```suggestion
However, it is still not the ideal thing to do, since we are relying on undefined `rpm-ostree` behavior, which may stop working some time in the future.
```

> Take a note that we are utilizing an OCI image, so this part can be ignored: `When OSTree commits get converted to OCI, /usr/etc/ becomes /etc/.`
> 
> Quoting Antheas:  
> :::note
> When OSTree commits get converted to OCI, /usr/etc/ becomes /etc/.
> Then during deployment, new /etc/ files are silently moved to /usr/etc/.
> 
> Relying on /usr/etc/ to place modified files in customized packages is a hack
> and relies on undefined behavior in the rpm-ostree source code that performs
> the path rewriting.
> :::

Reviewer: ```suggestion
Quoting Antheas:  
> When OSTree commits get converted to OCI, /usr/etc/ becomes /etc/. Then during deployment, new /etc/ files are silently moved to /usr/etc/.
> 
> Relying on /usr/etc/ to place modified files in customized packages is a hack and relies on undefined behavior in the rpm-ostree source code that performs the path rewriting.

(Take a note that we are utilizing an OCI image, so this part can be ignored: `When OSTree commits get converted to OCI, /usr/etc/ becomes /etc/.`)
```

As can be seen from the 1st paragraph of the note above, this is how `ostree` functions & this is why it's desired to transition to `/etc/`.

> `/etc/` files in build-time are copied to `/usr/etc/` during deployment, so local-users will note no changes.

Reviewer: ```suggestion
`/etc/` files in the image are copied to `/usr/etc/` during deployment, so local-users will experience no changes.
```

## How to transition

> 0. It is highly recommended that you follow & complete [The Caching Update](https://blue-build.org/blog/caching-update/) & [Updating the directory structure](https://blue-build.org/blog/config-files/) blog post updates
> 1. If you're working with a 1-to-1 filetree mapping structure, where you copy files & folders to the `destination` root (`/`) folder, then do the following (source-folder is the folder defined in `source` key inside files module recipe entry):
>   - Copy files from `files/source-folder/usr/etc/` to `files/source-folder/etc/`
> 2. If you copy files/folders outside of a 1-to-1 filetree mapping structure, then any files that go to `/usr/etc/` should be copied to go to `/etc/` according to your custom filetree mapping structure.
> 3. If you use [`script` module](https://blue-build.org/reference/modules/script/) & have any scripts which modify or download files to `/usr/etc/`, modify them to point out to `/etc/` files instead.
> 4. You're good to go! Commit your changes and wait to see that your custom image builds correctly.

Reviewer: ```suggestion
0. It is highly recommended that you follow & complete [The Caching Update](https://blue-build.org/blog/caching-update/) & [Updating the directory structure](https://blue-build.org/blog/config-files/) blog post updates
1. If you're working with a _1-to-1 filetree mapping_ structure, where you copy files & folders to the root (`/`) folder, then do the following (source-folder is the folder defined in `source` key inside files module recipe entry):
  - Move files from `files/system/usr/etc/` to `files/system/etc/` (or replace `system` with the name of the specific directory)
2. If you copy files to the image in other ways, then any files that go to `/usr/etc/` should be copied to go to `/etc/` according, including files copied by custom scripts.
3. You're good to go! Commit your changes and wait to see that your custom image builds correctly.
```
